File: amplify/functions/thumbnail-generator/handler.js
import { S3Client, GetObjectCommand, PutObjectCommand } from '@aws-sdk/client-s3';
import Jimp from 'jimp';

const s3Client = new S3Client({ region: process.env.AWS_REGION });

export const handler = async (event) => {
  console.log('Thumbnail generator triggered:', JSON.stringify(event, null, 2));
  
  try {
    // Parse the event - could be from API Gateway or S3 trigger
    let bucketName, objectKey;
    
    if (event.Records) {
      // S3 trigger event
      const record = event.Records[0];
      bucketName = record.s3.bucket.name;
      objectKey = decodeURIComponent(record.s3.object.key.replace(/\+/g, ' '));
    } else {
      // API Gateway event
      const body = typeof event.body === 'string' ? JSON.parse(event.body) : event.body;
      bucketName = body.bucketName || process.env.STORAGE_BUCKET_NAME;
      objectKey = body.objectKey;
    }
    
    console.log(`Processing: ${bucketName}/${objectKey}`);
    
    // Skip if this is already a thumbnail
    if (objectKey.includes('_thumb.')) {
      console.log('Skipping thumbnail file');
      return {
        statusCode: 200,
        body: JSON.stringify({ message: 'Skipped thumbnail file' })
      };
    }
    
    // Generate thumbnail key
    const thumbnailKey = generateThumbnailKey(objectKey);
    
    try {
      // Get original image from S3
      const getObjectParams = {
        Bucket: bucketName,
        Key: objectKey
      };
      
      const originalImage = await s3Client.send(new GetObjectCommand(getObjectParams));
      const imageBuffer = await streamToBuffer(originalImage.Body);
      
      // Generate thumbnail using Jimp
      const thumbnailBuffer = await generateThumbnailWithJimp(imageBuffer, 300, 80);
      
      // Upload thumbnail to S3
      const putObjectParams = {
        Bucket: bucketName,
        Key: thumbnailKey,
        Body: thumbnailBuffer,
        ContentType: 'image/jpeg',
        CacheControl: 'max-age=31536000', // 1 year cache
      };
      
      await s3Client.send(new PutObjectCommand(putObjectParams));
      
      console.log(`Thumbnail generated successfully: ${thumbnailKey}`);
      
      return {
        statusCode: 200,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Headers': 'Content-Type',
          'Access-Control-Allow-Methods': 'POST, OPTIONS'
        },
        body: JSON.stringify({
          message: 'Thumbnail generated successfully',
          originalKey: objectKey,
          thumbnailKey: thumbnailKey,
          thumbnailSize: thumbnailBuffer.length
        })
      };
      
    } catch (error) {
      console.error('Error processing image:', error);
      
      // Return success even if thumbnail generation fails
      // This prevents upload failures due to thumbnail issues
      return {
        statusCode: 200,
        headers: {
          'Access-Control-Allow-Origin': '*',
          'Access-Control-Allow-Headers': 'Content-Type',
          'Access-Control-Allow-Methods': 'POST, OPTIONS'
        },
        body: JSON.stringify({
          message: 'Original upload successful, thumbnail generation failed',
          error: error.message,
          originalKey: objectKey
        })
      };
    }
    
  } catch (error) {
    console.error('Lambda execution error:', error);
    
    return {
      statusCode: 500,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': 'Content-Type',
        'Access-Control-Allow-Methods': 'POST, OPTIONS'
      },
      body: JSON.stringify({
        error: 'Internal server error',
        message: error.message
      })
    };
  }
};

// Helper function to convert stream to buffer
async function streamToBuffer(stream) {
  const chunks = [];
  for await (const chunk of stream) {
    chunks.push(chunk);
  }
  return Buffer.concat(chunks);
}

// Helper function to generate thumbnail using Jimp
async function generateThumbnailWithJimp(imageBuffer, maxSize, quality) {
  try {
    // Load the image using Jimp
    const image = await Jimp.read(imageBuffer);
    
    // Resize with aspect ratio preservation
    // Jimp.AUTO maintains aspect ratio for the other dimension
    image.resize(maxSize, maxSize, Jimp.RESIZE_BEZIER);
    
    // Convert to JPEG with quality (0-100)
    const thumbnailBuffer = await image.quality(quality).getBufferAsync(Jimp.MIME_JPEG);
    
    return thumbnailBuffer;
  } catch (error) {
    console.error('Error generating thumbnail with Jimp:', error);
    throw error;
  }
}

// Helper function to generate thumbnail key
function generateThumbnailKey(originalKey) {
  const pathParts = originalKey.split('.');
  const extension = pathParts.pop();
  const basePath = pathParts.join('.');
  return `${basePath}_thumb.${extension}`;
}

File: amplify/functions/thumbnail-generator/handler.py
import boto3
import json
import os
import uuid
from urllib.parse import unquote_plus
from PIL import Image
import io

s3_client = boto3.client('s3')

def resize_image(image_data, max_size=300, quality=80):
    """Resize image maintaining aspect ratio"""
    print(f'[RESIZE] Starting resize_image - max_size={max_size}, quality={quality}')
    print(f'[RESIZE] Input image data size: {len(image_data)} bytes')

    with Image.open(io.BytesIO(image_data)) as image:
        print(f'[RESIZE] Original image loaded - size={image.size}, mode={image.mode}, format={image.format}')

        # Calculate thumbnail size maintaining aspect ratio
        original_size = image.size
        image.thumbnail((max_size, max_size), Image.LANCZOS)
        print(f'[RESIZE] Thumbnail created - original size={original_size}, new size={image.size}')

        # Save to bytes buffer as JPEG
        output_buffer = io.BytesIO()
        # Convert to RGB if image has transparency (RGBA)
        if image.mode in ('RGBA', 'LA'):
            print(f'[RESIZE] Converting from {image.mode} to RGB')
            background = Image.new('RGB', image.size, (255, 255, 255))
            background.paste(image, mask=image.split()[-1] if image.mode == 'RGBA' else None)
            image = background
            print(f'[RESIZE] Conversion complete - new mode={image.mode}')

        print(f'[RESIZE] Saving as JPEG - quality={quality}')
        image.save(output_buffer, format='JPEG', quality=quality, progressive=True)
        output_buffer.seek(0)
        output_data = output_buffer.getvalue()
        print(f'[RESIZE] Thumbnail saved - output size={len(output_data)} bytes')
        return output_data

def generate_thumbnail_key(original_key):
    """Generate thumbnail key with _thumb suffix"""
    path_parts = original_key.rsplit('.', 1)
    if len(path_parts) == 2:
        base_path, extension = path_parts
        return f"{base_path}_thumb.{extension}"
    return f"{original_key}_thumb"

def lambda_handler(event, context):
    """
    AWS Lambda handler for generating thumbnails from uploaded images.
    Following Amplify Gen 2 Python Lambda best practices.
    """
    print(f'Event: {json.dumps(event, default=str)}')
    
    try:
        # Parse GraphQL mutation event from Amplify
        args = event.get('arguments', {})
        object_key = args.get('objectKey')
        
        # Get bucket name from environment - set by Amplify automatically
        bucket_name = os.environ.get('AMPLIFY_STORAGE_BUCKET_NAME')
        
        if not bucket_name:
            raise Exception('Storage bucket name not found in environment variables')
        
        print(f'Processing: {bucket_name}/{object_key}')
        
        # Skip if this is already a thumbnail
        if '_thumb.' in object_key:
            print('Skipping thumbnail file')
            return {'message': 'Skipped thumbnail file'}
        
        # Generate thumbnail key
        thumbnail_key = generate_thumbnail_key(object_key)
        
        try:
            # Download original image from S3
            response = s3_client.get_object(Bucket=bucket_name, Key=object_key)
            image_data = response['Body'].read()
            
            # Generate thumbnail using PIL
            thumbnail_data = resize_image(image_data, max_size=600, quality=95)
            
            # Upload thumbnail to S3
            s3_client.put_object(
                Bucket=bucket_name,
                Key=thumbnail_key,
                Body=thumbnail_data,
                ContentType='image/jpeg',
                CacheControl='max-age=31536000'  # 1 year cache
            )
            
            print(f'Thumbnail generated successfully: {thumbnail_key}')
            
            return {
                'message': 'Thumbnail generated successfully',
                'originalKey': object_key,
                'thumbnailKey': thumbnail_key,
                'thumbnailSize': len(thumbnail_data)
            }
            
        except Exception as error:
            print(f'Error processing image: {str(error)}')
            
            # Return success even if thumbnail generation fails
            # This prevents upload failures due to thumbnail issues
            return {
                'message': 'Original upload successful, thumbnail generation failed',
                'error': str(error),
                'originalKey': object_key
            }
    
    except Exception as error:
        print(f'Lambda execution error: {str(error)}')
        raise Exception(f'Lambda execution error: {str(error)}')